### DISCUSSION TITLE
Boolean Expression Evaluator?

### DISCUSSION THREAD

**Discussion by test3211234** [zesterer/chumsky#582] [Q&A] [answered]
Hi, this crate looks really good, especially the name. I was wondering from the creator's perspective what the best way to go about making a VSCode when clause-like boolean expression evaluator with `&&`, `||`, `==`, `!=`, and potentially even `(...)`/`!(...)`. Like this: `"!something && some == !other || !something != another"`, using an external source to determine those variables like `something` and `some` and return true or false. Right now, from VSCode's source, I have this regex that finds any string/variable so that it can be searched and evaluted: `[a-zA-Z0-9_<>\-\./\\:\*\?\+\[\]\^,#@;%\$\p{L}-]+`. But I don't really know how I should go about this. Thanks for this library.

**Comment by zesterer** [comment]
You might find the [example pratt parser](https://docs.rs/chumsky/1.0.0-alpha.6/chumsky/trait.Parser.html#method.pratt) to be very close to what you want.

Despite being extremely terse, it's a fully-functioning expression evaluator. Of course, the 'proper' thing to do is to parse it into an AST and then evaluate the AST, but if you're just writing something quick, there's not necessarily a need to do that.

**Reply by test3211234** [reply]
Thanks. I don't really understand it though. Can you give me an example for my use case? And how do I get the alpha version? And why is this not "proper"? I'm writing a library.

**Reply by zesterer** [reply, marked as answer]
Parsing arithmetic operations like `+` is not really any different to parsing boolean operators like `==`, `&&`, etc. Each operator has precedence (indicated by the number given for each operator: see the pratt module docs for more information) and an associativity (left or right). Along with that, you provide a function that tells chumsky how it can combine the operator and operands together (usually you'd use this function to construct an AST node representing that operation, but in the example we just evaluate the result *immediately*).

You can read the [guide](https://docs.rs/chumsky/1.0.0-alpha.6/chumsky/guide/) to see how to use chumsky in a Rust project. The guide also covers key concepts in crate, and a little parser theory.

That said, your response implies that there might be a few gaps in your understanding, enough that a more general introduction to parsers and syntax trees might be useful. There are plenty of good explanations elsewhere on the internet about these things, and I'd strongly recommend building up a rudimentary understanding of the core ideas.

**Reply by test3211234** [reply]
Isn't it different if you can have strings inside single quotes like `something != 'in quotes'` and with not `!` before parentheses: `!(...)`? AND having to check if each string "variable" like `something` is true or false using a function like `check_value("something")`?

**Reply by zesterer** [reply]
Not really, that all just comes under the same conceptual umbrella. Parsing is not a magic or specific thing that only works with a particular set of incantations: you can write parsers for whatever kind of syntax you want, and chumsky can help you write them for a pretty sizeable subset of those.
